File: creditcarddesign/src/main/java/com/cooltechworks/creditcarddesign/CardSelector.java
package com.cooltechworks.creditcarddesign;

/**
 * Created by Harish on 01/01/16.
 */
public class CardSelector {

    public static final CardSelector VISA = new CardSelector(R.drawable.card_color_round_rect_purple, R.drawable.chip, R.drawable.chip_inner, android.R.color.transparent, R.drawable.ic_billing_visa_logo);
    public static final CardSelector MASTER = new CardSelector(R.drawable.card_color_round_rect_pink, R.drawable.chip_yellow, R.drawable.chip_yellow_inner, android.R.color.transparent, R.drawable.ic_billing_mastercard_logo);
    public static final CardSelector AMEX = new CardSelector(R.drawable.card_color_round_rect_green, android.R.color.transparent, android.R.color.transparent, R.drawable.img_amex_center_face, R.drawable.ic_billing_amex_logo1);
    public static final CardSelector DEFAULT = new CardSelector(R.drawable.card_color_round_rect_default, R.drawable.chip, R.drawable.chip_inner, android.R.color.transparent, android.R.color.transparent);
    private static final java.lang.String AMEX_PREFIX = "3";


    private int mResCardId;
    private int mResChipOuterId;
    private int mResChipInnerId;
    private int mResCenterImageId;
    private int mResLogoId;
    
    public CardSelector(int mDrawableCard, int mDrawableChipOuter, int mDrawableChipInner, int mDrawableCenterImage, int logoId) {
        this.mResCardId = mDrawableCard;
        this.mResChipOuterId = mDrawableChipOuter;
        this.mResChipInnerId = mDrawableChipInner;
        this.mResCenterImageId = mDrawableCenterImage;
        this.mResLogoId = logoId;
    }


    public int getResCardId() {
        return mResCardId;
    }

    public void setResCardId(int mResCardId) {
        this.mResCardId = mResCardId;
    }

    public int getResChipOuterId() {
        return mResChipOuterId;
    }

    public void setResChipOuterId(int mResChipOuterId) {
        this.mResChipOuterId = mResChipOuterId;
    }

    public int getResChipInnerId() {
        return mResChipInnerId;
    }

    public void setResChipInnerId(int mResChipInnerId) {
        this.mResChipInnerId = mResChipInnerId;
    }

    public int getResCenterImageId() {
        return mResCenterImageId;
    }

    public void setResCenterImageId(int mResCenterImageId) {
        this.mResCenterImageId = mResCenterImageId;
    }

    public int getResLogoId() {
        return mResLogoId;
    }

    public void setResLogoId(int mResLogoId) {
        this.mResLogoId = mResLogoId;
    }


    public static CardSelector selectCard(char cardFirstChar) {

        switch (cardFirstChar) {
            case '4':
                return VISA;
            case '5':
                return MASTER;
            case '3':
                return AMEX;
            default:
                return DEFAULT;
        }
    }

    public static CardSelector selectCard(String cardNumber) {

        if(cardNumber != null && cardNumber.length() >= 3) {
            CardSelector selector = selectCard(cardNumber.charAt(0));
            
            if(cardNumber.startsWith(AMEX_PREFIX)) {
                return AMEX;
            }

            if(selector != DEFAULT) {

                int[] drawables = { R.drawable.card_color_round_rect_brown, R.drawable.card_color_round_rect_green, R.drawable.card_color_round_rect_pink, R.drawable.card_color_round_rect_purple, R.drawable.card_color_round_rect_blue};
                int hash = cardNumber.substring(0,3).hashCode();

                if(hash<0) {
                    hash = hash * -1;
                }

                int index = hash % drawables.length;

                int chipIndex = hash % 3;
                int[] chipOuter = { R.drawable.chip, R.drawable.chip_yellow, android.R.color.transparent};
                int[] chipInner = { R.drawable.chip_inner, R.drawable.chip_yellow_inner,android.R.color.transparent};


                selector.setResCardId(drawables[index]);
                selector.setResChipOuterId(chipOuter[chipIndex]);
                selector.setResChipInnerId(chipInner[chipIndex]);

                return selector;
            }
        }
        return DEFAULT;
    }
}

File: creditcarddesign/src/main/java/com/cooltechworks/creditcarddesign/CardSelector.kt
package com.cooltechworks.creditcarddesign

/**
 * Created by Harish on 01/01/16.
 */
class CardSelector(var resCardId: Int, var resChipOuterId: Int, var resChipInnerId: Int, var resCenterImageId: Int, var resLogoId: Int, cvvLength: Int) {
    var cvvLength = CVV_LENGHT_DEFAULT

    init {
        this.cvvLength = cvvLength
    }

    companion object {

        val VISA = CardSelector(R.drawable.card_color_round_rect_purple, R.drawable.chip, R.drawable.chip_inner, android.R.color.transparent, R.drawable.ic_billing_visa_logo, CardSelector.CVV_LENGHT_DEFAULT)
        val MASTER = CardSelector(R.drawable.card_color_round_rect_pink, R.drawable.chip_yellow, R.drawable.chip_yellow_inner, android.R.color.transparent, R.drawable.ic_billing_mastercard_logo, CardSelector.CVV_LENGHT_DEFAULT)
        val AMEX = CardSelector(R.drawable.card_color_round_rect_green, android.R.color.transparent, android.R.color.transparent, R.drawable.img_amex_center_face, R.drawable.ic_billing_amex_logo1, CardSelector.CVV_LENGHT_AMEX)
        val DISCOVER = CardSelector(R.drawable.card_color_round_rect_brown, android.R.color.transparent, android.R.color.transparent, android.R.color.transparent, R.drawable.ic_billing_discover_logo, CardSelector.CVV_LENGHT_DEFAULT)
        val DEFAULT = CardSelector(R.drawable.card_color_round_rect_default, R.drawable.chip, R.drawable.chip_inner, android.R.color.transparent, android.R.color.transparent, CardSelector.CVV_LENGHT_DEFAULT)

        val CVV_LENGHT_DEFAULT = 3
        val CVV_LENGHT_AMEX = 4

        fun selectCardType(cardType: CreditCardUtils.CardType): CardSelector {
            when (cardType) {
                CreditCardUtils.CardType.AMEX_CARD -> return AMEX
                CreditCardUtils.CardType.DISCOVER_CARD -> return DISCOVER
                CreditCardUtils.CardType.MASTER_CARD -> return MASTER
                CreditCardUtils.CardType.VISA_CARD -> return VISA
                else -> return DEFAULT
            }
        }

        fun selectCard(cardNumber: String?): CardSelector {
            if (cardNumber != null && cardNumber.length >= 1) {
                val cardType = CreditCardUtils.selectCardType(cardNumber)
                val selector = selectCardType(cardType)

                if (selector !== DEFAULT && cardNumber.length >= 3) {
                    val drawables = intArrayOf(R.drawable.card_color_round_rect_brown, R.drawable.card_color_round_rect_green, R.drawable.card_color_round_rect_pink, R.drawable.card_color_round_rect_purple, R.drawable.card_color_round_rect_blue)
                    var hash = cardNumber.substring(0, 3).hashCode()

                    if (hash < 0) {
                        hash = hash * -1
                    }

                    val index = hash % drawables.size

                    val chipIndex = hash % 3
                    val chipOuter = intArrayOf(R.drawable.chip, R.drawable.chip_yellow, android.R.color.transparent)
                    val chipInner = intArrayOf(R.drawable.chip_inner, R.drawable.chip_yellow_inner, android.R.color.transparent)

                    selector.resCardId = drawables[index]
                    selector.resChipOuterId = chipOuter[chipIndex]
                    selector.resChipInnerId = chipInner[chipIndex]

                    return selector
                }
            }

            return DEFAULT
        }
    }
}
